lang/java/avro/src/main/java/org/apache/avro/data/Java8TimeConversions.java

```diff
@@ -0,0 +1,161 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.avro.data;
+
+import org.apache.avro.Conversion;
+import org.apache.avro.LogicalType;
+import org.apache.avro.Schema;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+
+public class Java8TimeConversions {
+  public static class DateConversion extends Conversion<LocalDate> {
+
+    @Override
+    public Class<LocalDate> getConvertedType() {
+      return LocalDate.class;
+    }
+
+    @Override
+    public String getLogicalTypeName() {
+      return "date";
+    }
+
+    @Override
+    public LocalDate fromInt(Integer daysFromEpoch, Schema schema, LogicalType type) {
+      return LocalDate.ofEpochDay(daysFromEpoch);
+    }
+
+    @Override
+    public Integer toInt(LocalDate date, Schema schema, LogicalType type) {
+      long epochDays = date.toEpochDay();
+
+      return Math.toIntExact(epochDays);
+    }
+  }
+
+  public static class TimeMillisConversion extends Conversion<LocalTime> {
+    @Override
+    public Class<LocalTime> getConvertedType() {
+      return LocalTime.class;
+    }
+
+    @Override
+    public String getLogicalTypeName() {
+      return "time-millis";
+    }
+
+    @Override
+    public LocalTime fromInt(Integer millisFromMidnight, Schema schema, LogicalType type) {
+      return LocalTime.ofNanoOfDay(1_000_000L * millisFromMidnight);
+    }
+
+    @Override
+    public Integer toInt(LocalTime time, Schema schema, LogicalType type) {
+      // this will always be in the integer range
+      return Math.toIntExact(time.toNanoOfDay() / 1_000_000);
+    }
+  }
+
+  public static class TimeMicrosConversion extends Conversion<LocalTime> {
+    @Override
+    public Class<LocalTime> getConvertedType() {
+      return LocalTime.class;
+    }
+
+    @Override
+    public String getLogicalTypeName() {
+      return "time-micros";
+    }
+
+    @Override
+    public LocalTime fromLong(Long microsFromMidnight, Schema schema, LogicalType type) {
+      return LocalTime.ofNanoOfDay(1_000L * microsFromMidnight);
+    }
+
+    @Override
+    public Long toLong(LocalTime time, Schema schema, LogicalType type) {
+      return time.toNanoOfDay() / 1_000L;
+    }
+  }
+
+  public static class TimestampMillisConversion extends Conversion<ZonedDateTime> {
```

Comment:
In the Avro spec, a timestamp is defined to be "independent of a particular time zone" (https://avro.apache.org/docs/1.8.2/spec.html#Logical+Types). As such would it be more correct to use java.time.Instant here?

Reply:
Fair enough I guess. I mimicked the Joda time classes used earlier and so instead of a `org.joda.time.DateTime` I used the `java.time.ZonedDateTime` which would be `java.time.` equivalent.